File: baekjoon/7576.py
# https://www.acmicpc.net/problem/7576
import queue

# board[N][M]
M, N = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(N)]

# M, N = 1000, 1000
# board = [[0]*M for _ in range(N)]
# board[0][0] = 1

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]


# BFS를 사용해야 한다.
def bfs():
    q = queue.Queue()

    # 1인 친구들을 모두 queue에 넣는다.
    want_size = 0
    flag = True
    for row in range(N):
        for col in range(M):
            if board[row][col] != -1:
                want_size += 1
            if board[row][col] == 1:
                q.put((row, col))
            elif board[row][col] == 0:
                flag = False

    if flag:
        return 0

    generation = 0
    cnt = 0
    while not q.empty():
        row, col = q.get()
        cnt += 1
        generation = board[row][col] - 1

        # row, col 주변에 0인 아이들을 칠해 뭐로? 자기꺼 +1로?
        for i in range(4):
            nx = row + dx[i]
            ny = col + dy[i]
            if 0 <= nx and nx < N and 0 <= ny and ny < M and board[nx][ny] == 0:
                board[nx][ny] = board[row][col] + 1
                q.put((nx, ny))

    if cnt != want_size:
        return -1

    return generation


print(bfs())

File: baekjoon/7576.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

typedef pair<int, int> ii;

int M,N;
int board[1000][1000];

int dx[] = {1, 0, -1, 0};
int dy[] = {0, 1, 0, -1};

int bfs() {
    queue<ii> q = queue<ii>();

    int want_size = 0;
    for (int row = 0; row < N; row++) {
        for (int col = 0; col < M; col++) {
            if (board[row][col] != -1)
                want_size++;
            if (board[row][col] == 1)
                q.push(ii(row, col));
        }
    }

    int gene = 0;
    int cnt = 0;

    while (!q.empty()) {
        int row = q.front().first;
        int col = q.front().second;
        q.pop();
        gene = board[row][col] - 1;
        cnt++;

        for (int i =0; i<4; i++) {
            int nx = row + dx[i];
            int ny = col + dy[i];
            if (0 <= nx && nx < N && 0 <= ny && ny < M && board[nx][ny] == 0) {
                board[nx][ny] = board[row][col] + 1;
                q.push(ii(nx,ny));
            }
        }
    }

    if (cnt != want_size) {
        return -1;
    }

    return gene;
}

int main(int argc, char const *argv[])
{
    cin >> M >> N;
    for (int row = 0; row < N; row++) {
        for (int col = 0; col < M; col++) {
            cin >> board[row][col];
        }
    }

    cout << bfs() << endl;

    return 0;
}
